### DISCUSSION TITLE
Set Filetype=Telekasten

### DISCUSSION THREAD

**Discussion by lambtho12** [nvim-telekasten/telekasten.nvim#30] [Ideas]
In the discussion related to Tags (https://github.com/renerocksai/telekasten.nvim/discussions/23), we briefly mentioned the possibility to set the filetype of telekasten files as filetype=telekasten.

In my opinion this is something that should really be considered, as it could improve enormously the interaction of telekasten with colorschemes and with other plugins (snippets comes directly to mind).

Regarding colorschemes, we could create highlight groups for all major elements. It will then become trivial for a maintainer of colorscheme to add proper support for telekasten (see [gruvbox-material](https://github.com/sainnhe/gruvbox-material/blob/fb27ccbd20cc1eda04e181f22c722977bdf9c934/colors/gruvbox-material.vim#L1305-L1325) or [Nord](https://github.com/arcticicestudio/nord-vim/blob/df497a7b396d32a721fa72bd9119db2bca3deed6/colors/nord.vim#L719-L735) for instance). I do not know how, but I guess in absence of proper config in the colorscheme it would fallback on classic markdown syntax.

Regarding snippets, most engine also work based on the filetype to activate some snippets or not. This would allow a user to set specific snippets that would only be usable in telekasten files and not general markdown files (such as README.md, etc).

I am not a big expert on the topic, so I can not think of any possible drawback. 😅

**Comment by renerocksai** [comment]
Thanks for insisting on this! It is a great enhancement to the telekasten!

I implemented it now and there is no more confusion about filetypes and highlighting. Setting the filetype to telekasten loads the syntax for telekasten. :smile: I removed the `markdown.vim` syntax from the plugin so existing markdown files will not receive telekasten syntax additions any more - unless they are opened by the plugin.

While I don't know if any colorscheme author would consider supporting telekasten.nvim, I think its syntax additions are
small enough so that one can easily add the highlights with colors picked from the desired theme oneself.

Snippets and other functionality: yeah! Separating telekasten notes from normal markdown files makes everything more flexible.

So far, I haven't seen any drawbacks. It just feels much cleaner now.

**Reply by Liquidmantis** [reply]
Is the telekasten filetype supposed to be automatic?  In my "home", `~/notes` for me, markdown files are not automatically setting as telekasten, I have to manually set it.  However, in the configured subdirectories, e.g. `home/daily`, the files are automatically set as telekasten.

Actually, with a bit more testing, it seems like the ft only automatically gets set if I use one of the telekasten functions to open a file.  If I "find" a file it gets set as telekasten, but if I open any markdown file in my configured home tree through any other method it's just set as markdown.

**Reply by renerocksai** [reply]
> Actually, with a bit more testing, it seems like the ft only automatically gets set if I use one of the telekasten functions to open a file. If I "find" a file it gets set as telekasten, but if I open any markdown file in my configured home tree through any other method it's just set as markdown.

That is the designed behavior. So markdown files like a README can remain markdown files and telekasten files get the telekasten treatment.

**Reply by renerocksai** [reply]
if you want telekasten always on you can use an autocommand like this

```vim
au filetype markdown set ft=telekasten
```

**Reply by Liquidmantis** [reply]
Okay, that’s what I figured.  I had just assumed it would be greedy in the set directories.  I’ll see if I can find a way to filter auto commands depending on cwd.  Thanks again!

It doesn't like my notes symlink, so I needed to use the resolved path, but this works for me:
`autocmd BufEnter /Users/liquid/Dropbox/documents/notes/*.md set filetype=telekasten`

Seems to work fine with the `daily`, etc. subdirectories, too.

`autocmd BufEnter */notes/*.md set filetype=telekasten` also works, and might be a useful solution for less generically named telekasten directories.

**Reply by renerocksai** [reply]
Telekasten does not get invoked when you open a markdown file. Markdown is a built-in filetype of neovim. Nvim knows what to do when it encounters one: set filetype to markdown, which in turn leads to set syntax to markdown, etc. With an autocmd you can undo all that. The reason it "worked" before this change was that telekasten issued an autocommand at `setup()` that would set the syntax of all markdown files to telekasten. So telekasten is much less greedy now. 

Ha!

Your autocmd is quite awesome, actually. Maybe telekasten can issue this command on `setup()` then. I am still thinking about that. With a magic autocmd like that built-in, one robs the users of defining their own if they don't like telekasten taking over automatically when they just happen to open a note. On the other hand: isn't that the point of the plugin? :thinking:

**Reply by renerocksai** [reply]
Soooo: why not make it an option? When in doubt, always make it an option :laughing:

I am thinking of the following: a config option `take_over_my_home` with default set to `true` would issue the autocmd for the configured `home`.

That would work!

**Reply by renerocksai** [reply]
I just pushed a9102eaed7942ca5cf5de48806d00c375eebfb00

This is both awesome and hilarious! :grin: : 

![image](https://user-images.githubusercontent.com/30892199/145654013-de7edb01-1d6a-4d5a-8211-1b9b75a83357.png)

Thanks for this awesome inspiration! :star_struck: :+1:

**Reply by Liquidmantis** [reply]
That's amazing!  It works great for me.

**Comment by Livingnight** [comment]
If you cast the filetypes to type telekasten, which lsp server do I use or how do I get another one, such as remark_ls, to recognize it? Currently when I create a new note none of my snippets work and no server attaches to the buffer.

**Reply by lambtho12** [reply]
Why do you need lsp? Is it only for snippets? If so, what snippet engine are you using?
I use [LuaSnip](https://github.com/L3MON4D3/LuaSnip). In my LuaSnip config, I have `ls.filetype_extend("telekasten", { "markdown" })`. It means that all markdown snippets are also applied to telekasten. I also have custom snippets only for telekasten files.

For syntax highlighting, it have [Treesitter](https://github.com/nvim-treesitter/nvim-treesitter) which seems to be working out of the box.

----
Anyway, if you have no use for telekasten filetype, you can disable it by setting `auto_set_filetype = false,` in telekasten's config. It will then inherit the filetype related to the extension (`*.md` by default)
